policy_agent/policy_agent.py

```diff
@@ -0,0 +1,698 @@
+from pycalico.datastore import DatastoreClient
+
+logger = logging.getLogger(__name__)
+util_logger = logging.getLogger(common.util.__name__)
+pycalico_logger = logging.getLogger(pycalico.__name__)
+
+_datastore_client = DatastoreClient()
+
+
+class PolicyAgent():
+
+    """
+    The Policy Agent is responsible for maintaining Watch Threads/Queues and internal resource lists.
+    """
+
+    def __init__(self):
+        self.watcher_queue = Queue.Queue()
+
+        self.namespaces = {}
+        self.services = {}
+        self.pods = {}
+        self.endpoints = {}
+
+        self.changed_namespaces = {}
+        self.changed_services = {}
+        self.changed_pods = {}
+        self.changed_endpoints = {}
+
+        self.namespace_watcher = None
+        self.service_watcher = None
+        self.pod_watcher = None
+        self.endpoints_watcher = None
+
+        self.init_lists()
+        self.init_threads()
+
+    def run(self):
+        """
+        PolicyAgent.run() is called at program init to spawn watch threads,
+        Loops to read responses from the _watcher Queue as they come in.
+        """
+        self.pod_watcher.start()
+        self.service_watcher.start()
+        self.namespace_watcher.start()
+        self.endpoints_watcher.start()
+
+        self.read_updates()
+
+    def init_lists(self):
+        """
+        Pull the initial list of existing resources and grabs the current Resource Version
+        for watcher use.
+        """
+        logger.info("Initializing Resource Lists")
+        initNamespaces = _get_api_list("namespaces")
+        initServices = _get_api_list("services")
+        initPods = _get_api_list("pods")
+        initEndpoints = _get_api_list("endpoints")
+
+        for ns in initNamespaces["items"]:
+            self.process_resource(command=CMD_ADDED,
+                                  kind=KIND_NAMESPACE,
+                                  resource_json=ns)
+
+        for svc in initServices["items"]:
+            self.process_resource(command=CMD_ADDED,
+                                  kind=KIND_SERVICE,
+                                  resource_json=svc)
+
+        for pod in initPods["items"]:
+            self.process_resource(command=CMD_ADDED,
+                                  kind=KIND_POD,
+                                  resource_json=pod)
+
+        for ep in initEndpoints["items"]:
+            self.process_resource(command=CMD_ADDED,
+                                  kind=KIND_ENDPOINTS,
+                                  resource_json=ep)
+
+        self.nsResourceVersion = initEndpoints["metadata"]["resourceVersion"]
+        self.svcResourceVersion = initServices["metadata"]["resourceVersion"]
+        self.podResourceVersion = initPods["metadata"]["resourceVersion"]
+        self.epResourceVersion = initNamespaces["metadata"]["resourceVersion"]
+
+        logger.info("Service Resource Version = %s" % self.svcResourceVersion)
```

Comment:
This log should be written in the form:

```python
logger.info("Service Resource Version = %s", self.svcResourceVersion)
```

I see a bunch of similar log statements that need correcting in the rest of this file.